File: src/1122.relative-sort-array.1217/solution.js
/**
 * @param {number[]} arr1
 * @param {number[]} arr2
 * @return {number[]}
 */
// 类似 791
var relativeSortArray = function (arr1, arr2) {
    // countMap统计在arr1中出现的arr2含有的元素的个数
    const countMap = {};
    for (let i = 0; i < arr2.length; i++) {
        countMap[arr2[i]] = 0;
    }
    // restArr存的是arr1中不在arr2的元素
    const restArr = [];
    for (let i = 0; i < arr1.length; i++) {
        if (countMap[arr1[i]] === undefined) {
            restArr.push(arr1[i]);
        } else {
            countMap[arr1[i]]++;
        }
    }
    const result = [];
    for (let i = 0; i < arr2.length; i++) {
        const count = countMap[arr2[i]];
        for (let j = 0; j < count; j++) {
            result.push(arr2[i]);
        }
    }
    return result.concat(restArr.sort((a, b) => a - b));
};

File: src/1122.relative-sort-array.1217/solution.java
class Solution {
    public int[] relativeSortArray(int[] arr1, int[] arr2) {
        Map<Integer, Long> countMap = Arrays.stream(arr1).boxed()
                .collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));
        List<Integer> rest = new ArrayList<>();
        List<Integer> result = new ArrayList<>();

        Set<Integer> set = new HashSet<>();

        for (int k : arr2) {
            long count = countMap.get(k);
            for (long j = 0; j < count; j++) {
                result.add(k);
            }
            set.add(k);
        }

        for (int k : arr1) {
            if (!set.contains(k)) {
                rest.add(k);
            }
        }

        rest.sort((a, b) -> a - b);

        result.addAll(rest);

        return result.stream().mapToInt(Integer::intValue).toArray();
    }
}